### ISSUE TITLE
WSGI and 2.13.28

### ISSUE THREAD

**Issue by jordic** [zopefoundation/Zope#464]
We can't still move to 4, but we want to deploy zope as wsgi, but we are hitting on some edge cases that makes the current implementation (on 2.13.28) not working:

1. After this change: https://github.com/zopefoundation/Zope/commit/9d0c6bf396fb45d58d6336e1c69f3487a763b0ae , you can't no longer commit a transaction inside a request (If we commit it, we loose the request (got closed throught the synchronizer). 

2. Exceptions from zExceptions are not properly captured, and they pop mostly as 500 return codes.

I saw that more ore less these problems had been solved on the 4.0 branch. Do you think it's possible to backport the wsgipublisher from 4.0 branch to 2.13.28? We can try to do this, but I'm just asking if seems faire, or it's something that will not work. Any thoughts?

**Comment by dataflake** [comment]
All effort is concentrated on Zope 4, first of all. I am sure none of the "usual suspects" will even begin considering this request.

Secondly, such a large code change is unacceptable for any 2.13 release because it's way more than just bug fix.

My advice would be to spend your time on making your code work with Zope 4 instead.

**Comment by icemac** [comment]
We used the WSGI publisher of Zope 2.13.28 until December 2018 in the union.cms project. We did not have the issues you described.

1. I'd consider it a bad practice to commit the transaction inside the request. Maybe a `savepoint` is enough.

2. We transformed the HTTP-500 exceptions in a WSGI middleware the correct HTTP status codes.

As the union.cms project no longer uses Zope 2 I personally see no need to port the changes in the WSGI publisher back to Zope 2. This backport would be way more than a bug fix. It changes the need for `repoze.tm` and `repoze.retry` and the way how errors are handled including error pages.

I fear the only way foreward is Zope 4.

**Comment by jordic** [comment]
So many thanks for the replys. We have a super old zope2 installation (lot's of sqlmethods, externalmethods... ), I know that the best path is to move to zope4 , but thouhgt it not will be so easy.. 
But anyways it's a step we will start working on. (Any advice on the path? How you have started moving?)

@icemac Agree with your point on commiting transactions, but our use case, is that at some point we had started doing long running requests, mostly batch jobs (don't ask me why...) interacting with a postgresdb.. and we have some critical issues here (long transactions/locks). We had started moving batch jobs outside, 
with it's own transaction machinery... but we still have some crazy (External methods), without tests, that are not so easy to port... After thinking a bit on the code, the main problem is that is using the zodb synchronizer to clean up the request.. there is another option here (just store the request on a thread local, and use a repoze_tm2 hook to clean it.. (this way, requests can be commited without loosing the our poor man state carrier (request))

For the middleware, it's what I had started doing, but later I thought that perhaps, if there are more people still deploying zope 2.13 (wsgi) makes sense to reuse the package.  So sorry, because I'm new on the community, I had started working on a zope environment one year and half ago... An nobody at my employer nows anything about zope (well, they just use the zmi to work and do things...) But we have to fix so many things to be able to live with the project...

Anyway, so many thanks for the replys and for the advices! We will try to contribute to the project on the near future.

**Comment by jugmac00** [comment]
@jordic I totally feel you!

I inherited quite a big Zope project myself - like 4 years ago, and also, there is nobody with Zope experience in my company.

Additionally, the Zope project is not my only task at the company.

Well, after thinking how to move ahead, I did three things:
- move to a hoster, which is experienced with the Zope stack ( https://flyingcircus.io )
- take part at the Zope migration sprints ( https://www.meetup.com/de-DE/Zope-Sprint/ )
- and finally outsourced the migration to a consulting company ( https://gocept.com/index_en.html ), which @icemac is part of

Feel free to drop me a line at @jugmac00 at Twitter if you want to swap experiences.

P.S.: There will be another Zope sprint in May in Halle - no matter how you decide on what to do, try to take part ( https://twitter.com/gocept/status/1082200091199524864 ).
